core/src/builtins.rs

```diff
@@ -0,0 +1,44 @@
+use solana_runtime::{
+    bank::{Builtin, Builtins, Entrypoint},
+    feature_set,
+};
+use solana_sdk::{genesis_config::ClusterType, pubkey::Pubkey};
+
+/// Builtin programs that are always available
+fn genesis_builtins(cluster_type: ClusterType) -> Vec<Builtin> {
+    let builtins = if cluster_type != ClusterType::MainnetBeta {
+        vec![
+            solana_bpf_loader_deprecated_program!(),
+            solana_bpf_loader_program!(),
+        ]
+    } else {
+        // Remove this `else` block and the `cluster_type` argument to this function once
+        // `feature_set::bpf_loader2_program::id()` is active on Mainnet Beta
+        vec![solana_bpf_loader_deprecated_program!()]
+    };
+
+    builtins
+        .into_iter()
+        .map(|b| Builtin::new(&b.0, b.1, Entrypoint::Loader(b.2)))
+        .collect()
+}
+
+/// Builtin programs activated dynamically by feature
+fn feature_builtins() -> Vec<(Builtin, Pubkey)> {
+    let builtins = vec![(
+        solana_bpf_loader_program!(),
+        feature_set::bpf_loader2_program::id(),
+    )];
+
+    builtins
+        .into_iter()
+        .map(|(b, p)| (Builtin::new(&b.0, b.1, Entrypoint::Loader(b.2)), p))
```

Comment:
`declare_builtin` should probably be a struct so we don't have reference it via tuple indexes.  If agreed we can do it a different PR

Reply:
Or just expand out the map to:
`.map(|((name, id, entrypoint), p)| (Builtin::new(&b.0, b.1, Entrypoint::Loader(b.2)), p))`

Reply:
> declare_builtin should probably be a struct so we don't have reference it via tuple indexes. If agreed we can do it a different PR

Yeah I agree that'll be much nicer.  I didn't want to take on that churn in this PR